### PR TITLE
Sample GFX Arch Detection

### PR DESCRIPTION
**Pull request by dlangbe**
Detect the GFX architecture when running samples supported across multiple architectures.
- Default to using parameters beneficial to GFX11 on perf_hgemm
- Add wave and block size checks to perf_hgemm

### CODE REVIEW

File: samples/common.hpp [hunk 1]
```diff
@@ -70,10 +70,28 @@ bool isF64Supported()
 
     std::string deviceName(mProps.gcnArchName);
 
-    return (deviceName.find("gfx90a") != std::string::npos);
+    return ((deviceName.find("gfx908") != std::string::npos)
+        || (deviceName.find("gfx90a") != std::string::npos)
+        || (deviceName.find("gfx940") != std::string::npos));
```
Comment (cgmillette) [inline samples/common.hpp:75]: Pls add gfx941, gfx942

File: samples/common.hpp [hunk 2]
```diff
@@ -83,8 +101,12 @@ bool isF32Supported()
 
     std::string deviceName(mProps.gcnArchName);
 
-    return (deviceName.find("gfx908") != std::string::npos)
-           || (deviceName.find("gfx90a") != std::string::npos);
+    return (deviceName.find("gfx90a") != std::string::npos);
```
Comment (cgmillette) [inline samples/common.hpp:104]: gfx940, gfx941, gfx942 will support f64

File: samples/perf_hgemm.cpp [hunk 3]
```diff
@@ -815,20 +861,20 @@ ROCWMMA_HOST void gemm_test(uint32_t m, uint32_t n, uint32_t k, float32_t alpha,
...
-                                                                                 matrixC.data(),
-                                                                                 matrixD_ref.data(),
-                                                                                 lda,
-                                                                                 ldb,
-                                                                                 ldc,
-                                                                                 ldd,
-                                                                                 alpha,
-                                                                                 beta);
+    std::vector<OutputT> matrixD_ref(m * n, std::numeric_limits<OutputT>::signaling_NaN());
+    gemm_cpu_h<InputT, OutputT, ComputeT, col_major, row_major, col_major>(m,
```
Comment (cgmillette) [inline samples/perf_hgemm.cpp:865]: Can probably update the layouts to their aliases too

File: samples/perf_hgemm.cpp [hunk 4]
```diff
@@ -667,11 +694,30 @@ ROCWMMA_HOST void gemm_test(uint32_t m, uint32_t n, uint32_t k, float32_t alpha,
     auto macroTileSize
         = rocwmma::make_coord2d(TBLOCK_X / warpSize * WARP_TILE_X, TBLOCK_Y * WARP_TILE_Y);
 
+    // Device check for supported block and wave sizes
+    if(isGfx11() && (ROCWMMA_M > 16 || ROCWMMA_N > 16))
```
Comment (cgmillette) [inline samples/perf_hgemm.cpp:698]: Only block size supported is 16 x 16 for gfx11. Can possibly adjust to check
isGFX11() && ((ROCWMMA_M != ROCWMMA_N) || (ROCWMMA_M != 16))

Can do a similar check for gfx9 (16x16 and 32x32):
isGfx9() && ((ROCWMMA_M != ROCWMMA_N) || (ROCWMMA_M != 16 && ROCWMMA_M != 32))

File: samples/perf_sgemm.cpp [hunk 5]
```diff
@@ -671,7 +671,7 @@ ROCWMMA_HOST void gemm_test(uint32_t m, uint32_t n, uint32_t k, float32_t alpha,
     if((m < get<0>(macroTileSize) || n < get<1>(macroTileSize) || k < ROCWMMA_K)
        || (m % ROCWMMA_M || n % ROCWMMA_N || k % ROCWMMA_K))
     {
-        std::cout << "Unsupported size!\n";
+        std::cout << "Unsupported matrix size!\n";
         return;
     }
```
Comment (cgmillette) [inline samples/perf_sgemm.cpp:677]: Need to do similar checks as above, except sgemm isn't valid on gfx11.
    
    if(isGfx11() )
    {
        std::cout << "Unsupported architecture!\n";
        return;
    }
   ...
   // check block and wave size for gfx9... etc.
   ...

File: samples/perf_dgemm.cpp [hunk 6]
```diff
@@ -671,7 +671,7 @@ ROCWMMA_HOST void gemm_test(uint32_t m, uint32_t n, uint32_t k, float64_t alpha,
     if((m < get<0>(macroTileSize) || n < get<1>(macroTileSize) || k < ROCWMMA_K)
        || (m % ROCWMMA_M || n % ROCWMMA_N || k % ROCWMMA_K))
     {
-        std::cout << "Unsupported size!\n";
+        std::cout << "Unsupported matrix size!\n";
         return;
     }
```
Comment (cgmillette) [inline samples/perf_dgemm.cpp:677]: Similar to sgemm. check block / wave sizes for gfx90a, 941, 942

**Review by cgmillette — CHANGES_REQUESTED** [review]
Let's add gfx941 and gfx942 support.
When checking block and wave sizes, we should enforce sizes found in the support table. This is because in the sample we might expect that folks will tinker with configuration values.

These checks can be enforced for all samples to prevent possible false positives

**Review by cgmillette — APPROVED** [review]
Looks great, thanks!

**Comment by cgmillette** [comment]
I'm going to restart the CI run. Had a failure on rocBLAS build earlier

**Comment by cgmillette** [comment]
Please also duplicate this PR against release-staging/rocm-rel-6.0

**Review by mkarunan — APPROVED** [review]

**Pull request merged**